lib/ide/src/transport/web.rs

```diff
@@ -0,0 +1,250 @@
+    /// obtaining it programmatically. Reported error codes are utterly
+    /// unreliable.
+    #[fail(display = "Failed to establish connection.")]
+    FailedToConnect,
+}
+
+/// Error that may occur when attempting to send the data over WebSocket
+/// transport.
+#[derive(Clone,Debug,Fail)]
+enum SendingError {
+    /// Calling `send` method has resulted in an JS exception.
+    #[fail(display = "Failed to send message. Exception: {:?}.", _0)]
+    FailedToSend(String),
+    /// The socket was already closed, even before attempting sending a message.
+    #[fail(display = "Failed to send message because socket state is {:?}.", _0)]
+    NotOpen(State),
+}
+
+
+
+// =============
+// === State ===
+// =============
+
+/// Describes the current state of WebSocket connection.
+#[derive(Clone,Copy,Debug,PartialEq)]
+pub enum State {
+    /// Socket has been created. The connection is not yet open.
+    Connecting,
+    /// The connection is open and ready to communicate.
+    Open,
+    /// The connection is in the process of closing.
+    Closing,
+    /// The connection is closed or couldn't be opened.
+    Closed,
+    /// Any other, unknown condition.
+    Unknown(u16),
+}
+
+impl State {
+    /// Returns current state of the given WebSocket.
+    pub fn query_ws(ws:&web_sys::WebSocket) -> State {
+        State::from_code(ws.ready_state())
+    }
+
+    /// Translates code returned by `WebSocket.readyState` into our enum.
+    /// cf https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState
+    pub fn from_code(code:u16) -> State {
+        match code {
+            web_sys::WebSocket::CONNECTING => State::Connecting,
+            web_sys::WebSocket::OPEN       => State::Open,
+            web_sys::WebSocket::CLOSING    => State::Closing,
+            web_sys::WebSocket::CLOSED     => State::Closed,
+            num                            => State::Unknown(num), // impossible
+        }
+    }
+}
+
+
+
+// =================
+// === WebSocket ===
+// =================
+
+/// Wrapper over JS `WebSocket` object and callbacks to its signals.
+#[derive(Debug)]
+pub struct WebSocket {
+    /// Handle to the JS `WebSocket` object.
+    pub ws         : web_sys::WebSocket,
+    /// Handle to a closure connected to `WebSocket.onmessage`.
+    pub on_message : ClosureStorage<MessageEvent>,
+    /// Handle to a closure connected to `WebSocket.onclose`.
+    pub on_close   : ClosureStorage<CloseEvent>,
+    /// Handle to a closure connected to `WebSocket.onopen`.
+    pub on_open    : ClosureStorage<Event>,
+    /// Handle to a closure connected to `WebSocket.onerror`.
+    pub on_error   : ClosureStorage<Event>,
+}
+
+impl WebSocket {
+    /// Wraps given WebSocket object.
+    pub fn new(ws:web_sys::WebSocket) -> WebSocket {
+        WebSocket {
+            ws,
+            on_message : default(),
+            on_close   : default(),
+            on_open    : default(),
+            on_error   : default(),
+        }
+    }
```

Comment:
I would put all the "storages" in a new struct. This way a lot of the code would be nicer, like you can then derive Default for that and change the `new` function to:

```rust
    pub fn new(ws:web_sys::WebSocket) -> WebSocket {
        let storage = default();
        WebSocket {ws,storage}
    }
```

Reply:
We would lose more on having write `.storages` than save here on `defaults`. 
Also, creating incidental data structures just to store things that can be defaulted does not feel nice to me.

Reply:
`We would lose more on having write .storages than save here on defaults` - I do not understand it. Actually writing `.storages` feels nice to me as it sorts the logic in code nicely. If you do not want to write it, you can always write deref for it. Anyway, I feel like grouping all storages into a common struct is actually a nice design.